core/shared/src/main/scala/zio/ZPool.scala

```diff
@@ -100,94 +117,251 @@ object ZPool {
-        else free.offer(attempted)
+        if (attempted.isFailure)
+          state.modify { case State(size, free) =>
+            if (size <= range.start)
+              allocate -> State(size, free + 1)
+            else
+              ZIO.unit -> State(size - 1, free)
+          }.flatten
+        else
+          state.update(state => state.copy(free = state.free + 1)) *>
+            items.offer(attempted) *>
+            track(attempted.result) *>
+            getAndShutdown.whenZIO(isShuttingDown.get)
 
       ZManaged.acquireReleaseWith(acquire)(release(_)).flatMap(_.toManaged)
     }
 
     /**
-     * Gets an item from the pool and shuts it down, returning `true` if this
-     * was successful, or `false` if the pool was empty.
+     * Begins pre-allocating pool entries based on minimum pool size.
      */
-    private def getAndShutdown(implicit trace: ZTraceElement): UIO[Boolean] =
-      size.get.map(_ > 0).tap { more =>
-        ZIO.when(more) {
-          free.take.flatMap { attempted =>
-            attempted.forEach(a => invalidated.update(_ - a)) *>
-              size.update(_ - 1) *>
-              attempted.finalizer
-          }
+    final def initialize(implicit trace: ZTraceElement): UIO[Unit] =
+      ZIO.replicateZIODiscard(range.start) {
+        ZIO.uninterruptibleMask { restore =>
+          state.modify { case State(size, free) =>
+            if (size < range.start)
+              (
+                for {
+                  reservation <- creator.reserve
+                  exit        <- restore(reservation.acquire).exit
+                  attempted   <- ZIO.succeed(Attempted(exit, reservation.release(Exit.succeed(()))))
+                  _           <- items.offer(attempted)
+                  _           <- track(attempted.result)
+                  _           <- getAndShutdown.whenZIO(isShuttingDown.get)
+                } yield attempted,
+                State(size + 1, free + 1)
+              )
+            else
+              ZIO.unit -> State(size, free)
+          }.flatten
         }
       }
 
+    def invalidate(item: A)(implicit trace: zio.ZTraceElement): UIO[Unit] =
+      invalidated.update(_ + item)
+
     /**
-     * Begins pre-allocating pool entries based on minimum pool size.
+     * Shrinks the pool down, but never to less than the minimum size.
      */
-    final def initialize(implicit trace: ZTraceElement): UIO[Unit] = ZIO.replicateZIODiscard(range.start)(allocate)
+    def shrink(implicit trace: ZTraceElement): UIO[Any] =
+      ZIO.uninterruptible {
+        state.modify { case State(size, free) =>
+          if (size > range.start && free > 0)
+            (
+              items.take.flatMap { attempted =>
+                attempted.forEach(a => invalidated.update(_ - a)) *>
+                  attempted.finalizer *>
+                  state.update(state => state.copy(size = state.size - 1))
+              },
+              State(size, free - 1)
```

Comment:
@adamgfraser Can I ask you why you're decrementing the state `size` this way: 

```scala
... >* state.update(state => state.copy(size = state.size - 1))
```

and not this way:

```scala
State(size - 1, free - 1)
```

So that if `attempted.forEach(a => invalidated.update(_ - a)) *> attempted.finalizer` fails the `size` is not decremented? 
`attempted.forEach(a => invalidated.update(_ - a)) *> attempted.finalizer` seems to be a `UIO` so I'm not sure to understand 🤔

Reply:
Yes we reduce `free` immediately and then `size` only after the resource has been finalized.